main.rb

```diff
@@ -1,102 +1,139 @@
-#The Enumarable project: Creating own methods
+# The Enumarable project: Creating own methods
+
+# frozen_string_literal: true
+
+# rubocop:disable Metrics/ModuleLength
 module Enumerable
+  def my_each
+    return to_enum unless block_given?
 
-    def my_each
-      raise LocalJumpError, 'You have not given a block to the method' unless block_given?
-      for el in self
-        yield el
-      end
+    tmp = is_a?(Range) ? to_a : self
+    i = 0
+    result = []
+    while i < tmp.size
+      result << yield(tmp[i])
+      i += 1
     end
+    result
+  end
 
-    def my_each_with_index
-        raise LocalJumpError, 'You have not given a block to the method' unless block_given?
-        i = 0
-        for el in self
-            yield i,el
-            i+=1 
-        end
-    end
+  def my_each_with_index
+    return to_enum :my_each_with_index unless block_given?
 
-    def my_select
-      raise LocalJumpError, 'You have not given a block to the method' unless block_given?
-      selected_items = []
-      my_each { |i| selected_items.push(i) if yield(i) }
-      selected_items
+    result = []
+    k = 0
+    my_each do |v|
+      result << yield(k, v)
+      k += 1
     end
+    result
+  end
 
-    def my_all?
-      raise LocalJumpError, 'You have not given a block to the method' unless block_given?
-      result = true
-      self.my_each do |e|
-        unless yield(e)
-          result = false
-          break
-        end
-      end
-      result
+  def my_select
+    return to_enum :my_select unless block_given?
+
+    selected_items = []
+    my_each { |i| selected_items.push(i) if yield(i) }
+    selected_items
+  end
+
+  # rubocop:disable  Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+  def my_all?(pattern = nil)
+    result = true
+    if block_given?
+      my_each { |i| result &= (yield i) }
+    elsif block_given? && !pattern.nil?
+      my_each { |i| result &= pattern == i }
+    else
+      my_each { |i| result &= i }
     end
+    result
+  end
```

Comment:
**[REQUIRED]**
* When a class is passed as an argument returns true if all of the collection is a member of such class.
```
Failure/Error: expect(array.my_all?(Integer)).to be array.all?(Integer)
     
       expected false
            got true
```
* When a Regex is passed as an argument returns true if all of the collection matches the Regex.
```
Failure/Error: expect(words.my_all?(/o/)).to be words.all?(/o/)
     
       expected false
            got true
```
* When a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern.
```
Failure/Error: expect(array.my_all?(3)).to be array.all?(3)
     
       expected false
            got true
```